components/alert/alert-toast.js

```diff
@@ -0,0 +1,183 @@
+	CLOSING: 'closing', // the close animation is running
+	PREOPENING: 'preopening', // a pause before running the opening animation because transitions won't run when changing from 'diplay: none' to 'display: block'
+	OPENING: 'opening', // the opening animation is running
+	OPEN: 'open' // the toast is open
+};
+
+class AlertToast extends LitElement {
+
+	static get properties() {
+		return {
+			buttonText: { type: String, attribute: 'button-text' },
+			hideCloseButton: { type: Boolean, attribute: 'hide-close-button' },
+			noAutoClose: { type: Boolean, attribute: 'no-auto-close' },
+			open: { type: Boolean, reflect: true },
+			subtext: { type: String },
+			type: { type: String, reflect: true },
+			_state: { type: String, reflect: true }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				display: block;
+			}
+
+			.d2l-alert-toast-container {
+				border-radius: 0.3rem;
+				bottom: 1.5rem;
+				box-shadow: 0 0.1rem 0.6rem 0 rgba(0,0,0,0.10);
+				display: none;
+				left: 0;
+				margin: 0 auto;
+				max-width: 600px;
+				position: fixed;
+				right: 0;
+				width: 100%;
+				z-index: 10000;
+			}
+
+			:host(:not([_state="closed"])) .d2l-alert-toast-container {
+				display: block;
+			}
+
+			:host([_state="opening"]) .d2l-alert-toast-container,
+			:host([_state="closing"]) .d2l-alert-toast-container {
+				transition-duration: 250ms;
+				transition-property: transform, opacity, visibility;
+				transition-timing-function: ease-in;
+			}
+
+			:host([_state="preopening"]) .d2l-alert-toast-container,
+			:host([_state="closing"]) .d2l-alert-toast-container {
+				opacity: 0;
+				transform: translateY(0.5rem);
+				visibility: hidden;
+			}
+
+			:host([_state="opening"]) .d2l-alert-toast-container {
+				opacity: 1;
+				transform: translateY(0);
+				visibility: visible;
+			}
+
+			.d2l-alert-toast {
+				animation: none;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this.hideCloseButton = false;
+		this.noAutoClose = false;
+		this.open = false;
+		this._state = states.CLOSED;
+	}
+
+	get _state() {
+		return this.__state;
+	}
+
+	set _state(val) {
+		const oldVal = this.__state;
+		if (oldVal !== val) {
+			this.__state = val;
+			this.requestUpdate('_state', oldVal);
+			this._stateChanged(val, oldVal);
+		}
+	}
+
+	get open() {
+		return this._open;
+	}
+
+	set open(val) {
+		const oldVal = this._open;
+		if (oldVal !== val) {
+			this._open = val;
+			this.requestUpdate('open', oldVal);
+			this._openChanged(val);
+		}
+	}
+
+	render() {
+		return html`
+			<div class="d2l-alert-toast-container" @transitionend=${this._onTransitionEnd}>
+				<d2l-alert class="d2l-alert-toast" type=${ifDefined(this.type)} @d2l-alert-closed=${this._onCloseClicked} button-text=${ifDefined(this.buttonText)} ?has-close-button=${!this.hideCloseButton} subtext=${ifDefined(this.subtext)}>
+					<slot></slot>
+				</d2l-alert>
+			</div>
+		`;
+	}
+
+	_onCloseClicked(e) {
+		e.target.removeAttribute('hidden');
```

Comment:
What's `e.target` in this context, the close button itself?

Reply:
It is the `d2l-alert`. It sets itself to `hidden` when the close button is clicked but this means the close transition isn't visible.

Reply:
Oh wow, so this is like undoing what the alert did to itself? So what we'd really want is a way to tell the alert to have a close button but to not actually do anything when it's clicked. Then the toast would still get the event but could handle the "closing" itself.

Reply:
Its definitely a bit weird, it feels a bit off that the alert actually closes itself when the button is clicked. There didn't feel like a great way to prevent it from closing itself given the current design of `d2l-alert` unless we add a `no-close` property but that felt bad so I carried this through from the Polymer version.

Reply:
I'm OK with keeping it as-is for now, but I think it's a good candidate for cleanup later.

Reply:
What if `d2l-alert` provided a slot, and the default content would be the button that exists.  The toast would put it's own button in that slot, which it wire up to.  That way `d2l-alert` can have it's default behavior, and we can override it for toast without dirtying up the API.  Thoughts??

Reply:
Another similar thought: what if when alert fired the `d2l-alert-close` event, it waited to see if someone called `preventDefault()` on it? If that was called, it wouldn't actually go through with the close. That's more what the spirit of `preventDefault` is for, ya?

Reply:
I like it, that feels like a very appropriate use of `preventDefault`!

Reply:
Yeah that would be another clean option.  So alert would be updated to listen for it's own `d2l-alert-close` event... and then in a `setTimeout` or `rAF` it would check `e.defaultPrevented` and proceed with closing if depending.

Reply:
I think it's even cleaner than that... events bubble up synchronously. So it could literally do this:
```
const e = new CustomEvent(
    'd2l-alert-closed', { bubbles: true, cancelable: true, composed: true }
);
this.dispatchEvent(e);
if (!e.defaultPrevented) {
    this.hidden = true;
}
```